fuel/utils/formats.py

```diff
@@ -1,8 +1,35 @@
 """Low-level utilities for reading a variety of source formats."""
+import codecs
+import gzip
 import tarfile
 import six
 
 
+def open_(filename, mode='r', encoding='utf-8'):
+    """Open a text file with UTF-8 and optional gzip compression.
+
+    This function is useful when dealing with text files.
+
+    Parameters
+    ----------
+    filename : str
+        The filename to read.
+    mode : str, optional
+        The mode with which to open the file. Defaults to `r`.
+    encoding : str, optional
+        The encoding to use (see the codecs documentation_ for supported
+        values). Defaults to `utf-8`.
+
+    .. _documentation:
+    https://docs.python.org/3/library/codecs.html#standard-encodings
+
+    """
+    if filename.endswith('.gz.'):
+        zf = gzip.open(filename, mode)
```

Comment:
Using `io.BufferedReader` (6a32dd44bd10eaf87ae7f41778f4942cd5ac339c) speeds things up a lot, it's now only 10% slower than the command line in my test (I'm comparing with `subprocess.Popen(['zcat', fn], stdout=subprocess.PIPE)`). For 4.5 million lines (670MB uncompressed, 270MB compressed) it takes `zcat` 5.5 seconds and `gzip.open` 6.4.

On my desktop the slowdown is about 2x compared to reading the raw text file (takes 2.9 seconds), but I have an SSD and lots of space; I figured that when reading from a network drive and when the text files take up a lot of space it might definitely be worth using `gzip` (keeping in mind that we have monolingual datasets that are 100GB *compressed*).